File: otto-mobile/otto-android-LEGACY-DO-NOT-OPEN/app/src/main/java/to/ottomot/driftd/routebuilder/engine/RouteRoadSnapper.kt
package to.ottomot.driftd.routebuilder.engine

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import to.ottomot.driftd.core.event.haversineMeters
import java.util.concurrent.TimeUnit

/**
 * Snaps route control points to roads via Mapbox Directions (ported from iOS [RouteRoadSnapper]).
 */
class RouteRoadSnapper(
    private val accessToken: String,
    private val httpClient: OkHttpClient = defaultHttpClient,
) {
    data class Result(
        val coordinates: List<RouteLatLng>,
        val distanceMeters: Double,
        val travelTimeSeconds: Double,
        val didSnapToRoad: Boolean,
        val turnManeuverCoordinates: List<RouteLatLng>,
    )

    suspend fun buildRoute(points: List<RouteLatLng>): Result =
        withContext(Dispatchers.IO) {
            buildRouteBlocking(points)
        }

    private fun buildRouteBlocking(points: List<RouteLatLng>): Result {
        if (points.size < 2) {
            return Result(
                coordinates = points,
                distanceMeters = 0.0,
                travelTimeSeconds = 0.0,
                didSnapToRoad = false,
                turnManeuverCoordinates = emptyList(),
            )
        }

        var allCoordinates = mutableListOf<RouteLatLng>()
        val allTurnCoordinates = mutableListOf<RouteLatLng>()
        var totalMeters = 0.0
        var totalSeconds = 0.0
        var didSnapAllSegments = true

        for (index in 0 until points.size - 1) {
            val segment = routeSegment(points[index], points[index + 1])
            if (allCoordinates.isEmpty()) {
                allCoordinates.addAll(segment.coordinates)
            } else {
                allCoordinates.addAll(segment.coordinates.drop(1))
            }
            allTurnCoordinates.addAll(segment.turnManeuverCoordinates)
            totalMeters += segment.distanceMeters
            totalSeconds += segment.travelTimeSeconds
            didSnapAllSegments = didSnapAllSegments && segment.didSnapToRoad
        }

        if (allCoordinates.isEmpty()) {
            allCoordinates = points.toMutableList()
        }

        return Result(
            coordinates = allCoordinates,
            distanceMeters = totalMeters,
            travelTimeSeconds = totalSeconds,
            didSnapToRoad = didSnapAllSegments,
            turnManeuverCoordinates = dedupeNearbyCoordinates(allTurnCoordinates, withinMeters = 40.0),
        )
    }

    private fun routeSegment(start: RouteLatLng, end: RouteLatLng): Result {
        if (!isValidCoordinate(start) || !isValidCoordinate(end)) {
            return fallbackSegment(start, end)
        }

        val coordPath = "${start.lng},${start.lat};${end.lng},${end.lat}"
        val url =
            "https://api.mapbox.com/directions/v5/mapbox/driving/$coordPath" +
                "?geometries=geojson&overview=full&steps=true&access_token=${accessToken.trim()}"

        return try {
            val request =
                Request.Builder()
                    .url(url)
                    .header("Accept", "application/json")
                    .get()
                    .build()
            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return fallbackSegment(start, end)
                val body = response.body?.string().orEmpty()
                if (body.isBlank()) return fallbackSegment(start, end)
                parseDirectionsResponse(body, start, end)
            }
        } catch (_: Exception) {
            fallbackSegment(start, end)
        }
    }

    private fun parseDirectionsResponse(
        body: String,
        start: RouteLatLng,
        end: RouteLatLng,
    ): Result {
        val root = JsonParser.parseString(body).asJsonObject
        val routes = root.getAsJsonArray("routes") ?: return fallbackSegment(start, end)
        if (routes.size() == 0) return fallbackSegment(start, end)

        val route = routes[0].asJsonObject
        val distanceMeters = route.get("distance")?.asDouble ?: 0.0
        val durationSeconds = route.get("duration")?.asDouble ?: 0.0
        val geometry = route.getAsJsonObject("geometry")
        val coordinates = parseGeoJsonCoordinates(geometry?.getAsJsonArray("coordinates"))
        val snappedCoordinates = if (coordinates.isEmpty()) listOf(start, end) else coordinates

        val turnCoordinates = mutableListOf<RouteLatLng>()
        val legs = route.getAsJsonArray("legs")
        if (legs != null) {
            for (legIndex in 0 until legs.size()) {
                val steps = legs[legIndex].asJsonObject.getAsJsonArray("steps") ?: continue
                turnCoordinates.addAll(turnManeuverCoordinates(steps))
            }
        }

        return Result(
            coordinates = snappedCoordinates,
            distanceMeters = distanceMeters,
            travelTimeSeconds = durationSeconds,
            didSnapToRoad = snappedCoordinates.isNotEmpty(),
            turnManeuverCoordinates = turnCoordinates,
        )
    }

    private fun turnManeuverCoordinates(steps: JsonArray): List<RouteLatLng> {
        val coordinates = mutableListOf<RouteLatLng>()
        for (index in 0 until steps.size()) {
            if (index == 0) continue
            val step = steps[index].asJsonObject
            val distance = step.get("distance")?.asDouble ?: 0.0
            if (distance < MINIMUM_TURN_STEP_METERS) continue

            val instructions =
                step.getAsJsonObject("maneuver")
                    ?.get("instruction")
                    ?.asString
                    ?.lowercase()
                    .orEmpty()
            if (instructions.contains("continue") && !instructions.contains("turn")) {
                continue
            }

            val stepCoords =
                parseGeoJsonCoordinates(
                    step.getAsJsonObject("geometry")?.getAsJsonArray("coordinates"),
                )
            val first = stepCoords.firstOrNull() ?: continue
            if (!isValidCoordinate(first)) continue
            coordinates.add(first)
        }
        return coordinates
    }

    private fun parseGeoJsonCoordinates(array: JsonArray?): List<RouteLatLng> {
        if (array == null || array.size() == 0) return emptyList()
        val coordinates = ArrayList<RouteLatLng>(array.size())
        for (index in 0 until array.size()) {
            val pair = array[index].asJsonArray
            if (pair.size() < 2) continue
            val lng = pair[0].asDouble
            val lat = pair[1].asDouble
            coordinates.add(lat to lng)
        }
        return coordinates
    }

    private fun fallbackSegment(start: RouteLatLng, end: RouteLatLng): Result =
        Result(
            coordinates = listOf(start, end),
            distanceMeters = 0.0,
            travelTimeSeconds = 0.0,
            didSnapToRoad = false,
            turnManeuverCoordinates = emptyList(),
        )

    private fun dedupeNearbyCoordinates(
        coordinates: List<RouteLatLng>,
        withinMeters: Double,
    ): List<RouteLatLng> {
        val deduped = mutableListOf<RouteLatLng>()
        for (coordinate in coordinates) {
            val isDuplicate =
                deduped.any { existing ->
                    haversineMeters(
                        coordinate.lat,
                        coordinate.lng,
                        existing.lat,
                        existing.lng,
                    ) <= withinMeters
                }
            if (!isDuplicate) {
                deduped.add(coordinate)
            }
        }
        return deduped
    }

    private fun isValidCoordinate(coordinate: RouteLatLng): Boolean {
        val lat = coordinate.lat
        val lng = coordinate.lng
        return lat.isFinite() && lng.isFinite() && lat in -90.0..90.0 && lng in -180.0..180.0
    }

    private companion object {
        private const val MINIMUM_TURN_STEP_METERS = 25.0

        private val defaultHttpClient: OkHttpClient =
            OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build()
    }
}

File: otto-mobile/otto-mobile/RouteRoadSnapper.swift
import CoreLocation
import MapKit

enum RouteRoadSnapper {
    struct Result {
        let coordinates: [CLLocationCoordinate2D]
        let distanceMeters: Double
        let travelTimeSeconds: TimeInterval
        let didSnapToRoad: Bool
        let turnManeuverCoordinates: [CLLocationCoordinate2D]
    }

    private static let minimumTurnStepMeters = 25.0

    static func buildRoute(for points: [CLLocationCoordinate2D]) async -> Result {
        guard points.count >= 2 else {
            return Result(
                coordinates: points,
                distanceMeters: 0,
                travelTimeSeconds: 0,
                didSnapToRoad: false,
                turnManeuverCoordinates: []
            )
        }

        var allCoordinates: [CLLocationCoordinate2D] = []
        var allTurnCoordinates: [CLLocationCoordinate2D] = []
        var totalMeters: Double = 0
        var totalSeconds: TimeInterval = 0
        var didSnapAllSegments = true

        for pair in zip(points, points.dropFirst()) {
            let segment = await routeSegment(from: pair.0, to: pair.1)
            if allCoordinates.isEmpty {
                allCoordinates.append(contentsOf: segment.coordinates)
            } else {
                allCoordinates.append(contentsOf: segment.coordinates.dropFirst())
            }
            allTurnCoordinates.append(contentsOf: segment.turnManeuverCoordinates)
            totalMeters += segment.distanceMeters
            totalSeconds += segment.travelTimeSeconds
            didSnapAllSegments = didSnapAllSegments && segment.didSnapToRoad
        }

        return Result(
            coordinates: allCoordinates.isEmpty ? points : allCoordinates,
            distanceMeters: totalMeters,
            travelTimeSeconds: totalSeconds,
            didSnapToRoad: didSnapAllSegments,
            turnManeuverCoordinates: dedupeNearbyCoordinates(allTurnCoordinates, withinMeters: 40)
        )
    }

    private static func mapItem(for coordinate: CLLocationCoordinate2D) -> MKMapItem? {
        guard CLLocationCoordinate2DIsValid(coordinate) else { return nil }
        return MKMapItem(
            location: CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude),
            address: nil
        )
    }

    private static func routeSegment(
        from start: CLLocationCoordinate2D,
        to end: CLLocationCoordinate2D
    ) async -> Result {
        guard let sourceItem = mapItem(for: start),
              let destinationItem = mapItem(for: end)
        else {
            return Result(
                coordinates: [start, end],
                distanceMeters: 0,
                travelTimeSeconds: 0,
                didSnapToRoad: false,
                turnManeuverCoordinates: []
            )
        }

        let request = MKDirections.Request()
        request.source = sourceItem
        request.destination = destinationItem
        request.transportType = .automobile
        request.requestsAlternateRoutes = false

        do {
            let response = try await MKDirections(request: request).calculate()
            guard let route = response.routes.first else {
                return Result(
                    coordinates: [start, end],
                    distanceMeters: 0,
                    travelTimeSeconds: 0,
                    didSnapToRoad: false,
                    turnManeuverCoordinates: []
                )
            }
            let polyline = route.polyline
            var coords = Array(repeating: CLLocationCoordinate2D(), count: polyline.pointCount)
            polyline.getCoordinates(&coords, range: NSRange(location: 0, length: polyline.pointCount))
            return Result(
                coordinates: coords.isEmpty ? [start, end] : coords,
                distanceMeters: route.distance,
                travelTimeSeconds: route.expectedTravelTime,
                didSnapToRoad: !coords.isEmpty,
                turnManeuverCoordinates: turnManeuverCoordinates(from: route)
            )
        } catch {
            return Result(
                coordinates: [start, end],
                distanceMeters: 0,
                travelTimeSeconds: 0,
                didSnapToRoad: false,
                turnManeuverCoordinates: []
            )
        }
    }

    private static func turnManeuverCoordinates(from route: MKRoute) -> [CLLocationCoordinate2D] {
        var coordinates: [CLLocationCoordinate2D] = []
        for (index, step) in route.steps.enumerated() {
            guard index > 0 else { continue }
            guard step.distance >= minimumTurnStepMeters else { continue }
            let instructions = step.instructions.lowercased()
            if instructions.contains("continue"), !instructions.contains("turn") {
                continue
            }
            guard step.polyline.pointCount > 0 else { continue }
            var coordinate = CLLocationCoordinate2D()
            step.polyline.getCoordinates(&coordinate, range: NSRange(location: 0, length: 1))
            guard CLLocationCoordinate2DIsValid(coordinate) else { continue }
            coordinates.append(coordinate)
        }
        return coordinates
    }

    private static func dedupeNearbyCoordinates(
        _ coordinates: [CLLocationCoordinate2D],
        withinMeters threshold: CLLocationDistance
    ) -> [CLLocationCoordinate2D] {
        var deduped: [CLLocationCoordinate2D] = []
        for coordinate in coordinates {
            let location = CLLocation(latitude: coordinate.latitude, longitude: coordinate.longitude)
            let isDuplicate = deduped.contains { existing in
                let existingLocation = CLLocation(latitude: existing.latitude, longitude: existing.longitude)
                return location.distance(from: existingLocation) <= threshold
            }
            if !isDuplicate {
                deduped.append(coordinate)
            }
        }
        return deduped
    }
}

// MARK: - Mapbox Directions (turn-by-turn route drive)

enum TurnByTurnRouteServiceError: LocalizedError {
    case insufficientWaypoints
    case missingAccessToken
    case invalidResponse
    case noRoutes
    case httpStatus(Int)

    var errorDescription: String? {
        switch self {
        case .insufficientWaypoints:
            return "At least two route points are required."
        case .missingAccessToken:
            return "Mapbox access token is missing."
        case .invalidResponse:
            return "Directions response was invalid."
        case .noRoutes:
            return "No driving route was found."
        case .httpStatus(let code):
            return "Directions request failed (\(code))."
        }
    }
}

struct TurnByTurnRouteService: NavigationRouteProviding {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func fetchRoute(waypoints: [CLLocationCoordinate2D]) async throws -> NavigationRoute {
        guard waypoints.count >= 2 else { throw TurnByTurnRouteServiceError.insufficientWaypoints }
        guard let token = MapboxAccessToken.current, !token.isEmpty else {
            throw TurnByTurnRouteServiceError.missingAccessToken
        }

        let coordinatePath = waypoints
            .map { String(format: "%.6f,%.6f", $0.longitude, $0.latitude) }
            .joined(separator: ";")
        var components = URLComponents(string: "https://api.mapbox.com/directions/v5/mapbox/driving/\(coordinatePath)")!
        components.queryItems = [
            URLQueryItem(name: "geometries", value: "geojson"),
            URLQueryItem(name: "overview", value: "full"),
            URLQueryItem(name: "steps", value: "true"),
            URLQueryItem(name: "voice_instructions", value: "true"),
            URLQueryItem(name: "banner_instructions", value: "true"),
            URLQueryItem(name: "voice_units", value: "imperial"),
            URLQueryItem(name: "annotations", value: "distance,duration"),
            URLQueryItem(name: "access_token", value: token),
        ]

        guard let url = components.url else { throw TurnByTurnRouteServiceError.invalidResponse }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse, !(200...299).contains(http.statusCode) {
            throw TurnByTurnRouteServiceError.httpStatus(http.statusCode)
        }
        return try Self.parseResponse(data: data, fallbackFinish: waypoints.last!)
    }

    static func parseResponse(data: Data, fallbackFinish: CLLocationCoordinate2D) throws -> NavigationRoute {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let routes = json["routes"] as? [[String: Any]],
              let route = routes.first else {
            throw TurnByTurnRouteServiceError.noRoutes
        }

        let totalDistance = route["distance"] as? Double ?? 0
        let totalDuration = route["duration"] as? Double ?? 0
        let geometry = route["geometry"] as? [String: Any]
        let coordinates = parseGeoJSONCoordinates(geometry?["coordinates"] as? [[Double]])
        guard coordinates.count >= 2 else { throw TurnByTurnRouteServiceError.invalidResponse }

        let polylineIndex = RoutePolylineIndex(lineCoordinates: coordinates)
        let legsJSON = route["legs"] as? [[String: Any]] ?? []
        var legs: [NavigationLeg] = []
        for legJSON in legsJSON {
            let legDistance = legJSON["distance"] as? Double ?? 0
            let legDuration = legJSON["duration"] as? Double ?? 0
            let stepsJSON = legJSON["steps"] as? [[String: Any]] ?? []
            let steps = stepsJSON.compactMap { parseStep($0, polylineIndex: polylineIndex) }
            legs.append(NavigationLeg(steps: steps, distanceMeters: legDistance, durationSeconds: legDuration))
        }
        legs = NavigationInstructionLabeling.relabelLegsForStopPoints(legs)

        return NavigationRoute(
            coordinates: coordinates,
            legs: legs,
            totalDistanceMeters: totalDistance,
            totalDurationSeconds: totalDuration,
            finishCoordinate: coordinates.last ?? fallbackFinish
        )
    }

    private static func parseStep(
        _ stepJSON: [String: Any],
        polylineIndex: RoutePolylineIndex
    ) -> NavigationStep? {
        let distance = stepJSON["distance"] as? Double ?? 0
        let duration = stepJSON["duration"] as? Double ?? 0
        let name = stepJSON["name"] as? String
        let maneuverJSON = stepJSON["maneuver"] as? [String: Any] ?? [:]
        let maneuverType = maneuverJSON["type"] as? String ?? "turn"
        let maneuverModifier = maneuverJSON["modifier"] as? String
        let maneuverInstruction = maneuverJSON["instruction"] as? String ?? "Continue"
        let geometry = stepJSON["geometry"] as? [String: Any]
        let geometryCoordinates = parseGeoJSONCoordinates(geometry?["coordinates"] as? [[Double]])

        let maneuverCoordinate: CLLocationCoordinate2D
        if let location = maneuverJSON["location"] as? [Double], location.count >= 2 {
            maneuverCoordinate = CLLocationCoordinate2D(latitude: location[1], longitude: location[0])
        } else if let first = geometryCoordinates.first {
            maneuverCoordinate = first
        } else {
            return nil
        }

        let maneuverArcLength = polylineIndex.projectOntoPolyline(maneuverCoordinate)?.arcLengthMeters ?? 0
        let voiceJSON = stepJSON["voiceInstructions"] as? [[String: Any]] ?? []
        let voiceInstructions = voiceJSON.compactMap { entry -> NavigationVoiceInstruction? in
            guard let announcement = entry["announcement"] as? String else { return nil }
            return NavigationVoiceInstruction(
                distanceAlongStepMeters: entry["distanceAlongGeometry"] as? Double ?? 0,
                announcement: NavigationSSMLCleaner.plainText(from: announcement)
            )
        }
        let instruction = NavigationSSMLCleaner.plainText(from: maneuverInstruction)
        return NavigationStep(
            instruction: instruction,
            name: name?.isEmpty == false ? name : nil,
            distanceMeters: distance,
            durationSeconds: duration,
            maneuver: NavigationManeuver(type: maneuverType, modifier: maneuverModifier, instruction: instruction),
            maneuverCoordinate: maneuverCoordinate,
            voiceInstructions: voiceInstructions,
            geometryCoordinates: geometryCoordinates,
            maneuverArcLengthMeters: maneuverArcLength
        )
    }

    private static func parseGeoJSONCoordinates(_ raw: [[Double]]?) -> [CLLocationCoordinate2D] {
        (raw ?? []).compactMap { pair in
            guard pair.count >= 2 else { return nil }
            let coordinate = CLLocationCoordinate2D(latitude: pair[1], longitude: pair[0])
            guard CLLocationCoordinate2DIsValid(coordinate), pair[0].isFinite, pair[1].isFinite else { return nil }
            return coordinate
        }
    }
}
